### ISSUE TITLE
`token.sent_start` cannot be set, even if doc is not parsed (but only tagged)

### ISSUE THREAD

**Issue by peter-exos** [explosion/spaCy#9782]
## How to reproduce the behaviour

A minimal example to illustrate the problem:
```
import en_core_web_sm
nlp = en_core_web_sm.load(disable=['parser'])

text = 'Big  news. Great deals.'
doc = nlp(text)
print([(t.i, t.text) for t in doc])

doc[4].sent_start = True
```

This breaks with the following output and error message
```
[(0, 'Big'), (1, ' '), (2, 'news'), (3, '.'), (4, 'Great'), (5, 'deals'), (6, '.')]

Traceback (most recent call last):
  File "...", line 8, in <...>
    doc[4].sent_start = True
  File "spacy/tokens/token.pyx", line 479, in spacy.tokens.token.Token.sent_start.__set__
  File "spacy/tokens/token.pyx", line 500, in spacy.tokens.token.Token.is_sent_start.__set__
ValueError: [E043] Refusing to write to token.sent_start if its document is parsed, because this may cause inconsistent state.
```
Without the additional whitespace, no exception is thrown.

The source of this specific problem seems to be the `attribute_ruler` packaged with the model, which assigns the dependency label `dep` to space tokens. The relevant rule is
```
 {'attrs': {'DEP': 'dep', 'MORPH': '_', 'POS': 'SPACE', 'TAG': '_SP'},
  'index': 0,
  'patterns': [[{'IS_SPACE': True}]]}
```

### Our use case

The above issue arises with our own rule-based sentence splitter if the pipeline also includes the tagger and attribute_ruler (we never use the parser, because it performs poorly in our domain, but the tagger can be useful, although we don't always need it downstream). 

The issue does not arise with spacy's rule-based `sentencizer`, because in its latest versions it is written in Cython and manipulates the underlying C struct directly. But our sentence splitter is written in pure Python and we thinks it works better in some edge cases (at least in our domain): it is actually based on an older version of the `sentencizer`, and we'd be more than happy to share our minor changes, but we'd also like to be able to make more changes to our sentence splitter without adding Cython (and its complications) to our deployment process.

### General discussion

The specific problem reported above can easily be fixed by removing `'DEP': 'dep'` from the `attrs` in the rule listed above. But I think it might be worth discussing the above exception more generally.

I can see the use of the above exception, but at least in my personal opinion it goes a bit against the spirit of empowering developers working with spacy: there may actually be use cases where additional rule-based sentence splitting may be desired after the parser (e.g. we may want to be certain that sentence boundaries are assigned at specific positions, and be fine if they are inconsistent with the overall parse, maybe because we only care about the dependency labels, or because we only care about the parse of a subtree).

So maybe there is a way to add an `overwrite` flag to allow developers to express that they want to overwrite previous sentence boundary annotations. I really liked the addition of this option to trainable components.

## Your Environment

- **spaCy version:** 3.2.0
- **Platform:** Darwin-18.7.0-x86_64-i386-64bit
- **Python version:** 3.7.3
- **Pipelines:** en_core_web_sm (3.2.0)

**Comment by databill86** [comment]
Hello, 
I've successfully reproduced this problem with `fr_core_news_lg-3.2.0` too. 
Although, everything work as expected with `fr_core_news_lg-3.1.0`, even with texts containing a sequence of more than one space.

**Comment by adrianeboyd** [comment]
This particular issue looks like an unintended side effect of changes to the attribute ruler rules. The current trained pipelines don't handle whitespace tokens well (due to a lack of training data on our end, not any technical reasons; we should add data augmentation for this in the future) and users complained that they were seeing nonsense dependency annotation like `nsubj` on whitespace, so we extended this rule. We can split it up into two rules for the next model release so that `DEP` is only set if there's an existing dep label.

It also points to a minor bug in the attribute ruler, since the sentence boundaries aren't synced correctly after `DEP` is set. This won't affect your problem, but it does mean that the doc might not be in a valid state after the attribute modifies `DEP`, so we should take another look at that.

But the constraint on setting sentence boundaries when dependency parses are present is an intentional part of the design, not a desire to thwart users. The `Doc` and `Token` classes have been developed with the assumption that sentence boundaries correspond to dependency parses if dependency parses are present, and a number of properties and methods could crash or produce incorrect results if this isn't the case. There's no easy way to handle partial dependency annotation, so as soon there are any dependencies, this constraint kicks in so that the sentence boundary annotation always stays in sync with the parses.

Basically, once you have parses, you can modify the parses to change the sentence boundaries, but you can't modify the sentence boundaries with `is_sent_start` anymore. If you reset all the dependency annotation, you should be able to set `is_sent_start` again. (In general, do use `is_sent_start` and not `sent_start`, which is deprecated.)

(As a side note, I wouldn't recommend resetting dependency annotation in python because it will be very slow, since it recalculates the sentence boundaries every time you modify a head. If you are going to do this for some reason, I think an attribute ruler should handle the cython side to do it fairly efficiently, so something like:

```python
 {'attrs': {'DEP': '', 'HEAD': 0, 'SENT_START': 0},
  'index': 0,
  'patterns': [[{}]]}
```

And then you can add the sentence boundaries that you want. If we fix the attribute ruler `DEP`/sents syncing bug mentioned above, then the state after this would probably be that each word thinks it's its own sentence rather than that they're unset, though you could still modify them directly afterwards in python.)

**Comment by peter-exos** [comment]
Thank you @adrianeboyd, that is very helpful to clarify the logic behind that exception.

Following that logic I would claim that you only want to raise that exception if the Doc has a complete parse: if there are only a few Tokens with DEP annotations (as assigned e.g. by the whitespace rule) there is no clear correspondence between sentence boundaries and dependency parses.

So would it be possible to replace https://github.com/explosion/spaCy/blob/72f7f4e68a5076a87dd9402812bfb72e479237ed/spacy/tokens/token.pyx#L499
with the following?
```
if self.doc.has_annotation("DEP", require_complete=True):
```
This should still throw an exception if sentence boundaries are manually set after the parser, but not if an attribute_ruler sets a few DEP labels before a custom sentence splitter.

**Comment by adrianeboyd** [comment]
I think the right fix here is to fix the attribute ruler rule instead.

There's no good line to draw between a little dependency annotation and a lot of dependency annotation since any new sentence boundary could potentially come in the middle of some dependency parse. And if some cases are permitted and some aren't, I think it will be confusing for users, too.

The `Doc` doesn't really deal with partial dependency annotation that well, so sentence boundaries may still be unexpected in some cases, and setting dependencies incrementally may also not work cleanly. I just tried some cases where I set the parses incrementally in python and the sentence boundaries weren't correct in the end. Serializing and reloading the fully-parsed doc had the expected sentence starts, but not just after setting everything incrementally. The code that deals with adjusting all the token properties when parses are modified is tricky, and I think trying to make it efficient for fully-parsed docs has led to cases that aren't handled correctly for partial parses. But these kinds of bugs don't really come up in practice too much, since most people just work with full parses from the parser, and I think the test suite also doesn't have enough coverage for cases with partial parses.
